Fix build_manpage.py with newer markdown versions

emanuele6 opened the pull request:
Yet another python packages problem. =)

The `markdown` package changed its extensions API slightly, so bumping it broke `build_manpage.py`, but not `build_mantests.py` and `build_website.py` since only the former was registering extensions.

Luckly it seems to be a simple fix.

<s>A CI job should probably make sure that the man page can be built.</s>I added a job that builds `tests/man.test` and `jq.1.prebuilt`, and makes sure that `ŧests/man.test` is up to date (<s>not `jq.1.prebuilt` since it contains a date</s> also `jq.1.prebuilt`).

emanuele6 committed:
Fix build_manpage.py with newer markdown versions

The markdown python changed its API for markdown extensions.

Bumping the markdown package version in previous commits broke
docs/build_manpage.py since it was using registering an extension with
the old API.

emanuele6 committed:
Add CI workflow to make sure the man page and man.test can be built

And that man.test is up to date.

itchyny commented on .github/workflows/manpage.yml:
```diff
@@ -0,0 +1,30 @@
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
```
The git history is unnecessary.

itchyny commented on .github/workflows/manpage.yml:
```diff
@@ -0,0 +1,29 @@
+  push:
+    paths:
+      - 'docs/**'
+      - '.github/workflows/manpage.yml'
```
Add the target files here to make WF fail on unexpected edits (e.x. add a test to man.test but not to manual.yml).

itchyny commented on .github/workflows/manpage.yml:
```diff
@@ -0,0 +1,29 @@
+          rm jq.1.prebuilt tests/man.test
+          make jq.1.prebuilt tests/man.test
+      - name: Make sure tests are up to date
+        run: git diff tests/man.test
```
Drop `--quiet` and use `--exit-code` instead, to make it understandable what happened on diff failure.

emanuele6 replied:
Oh, I already dropped `--quiet`, but I didn't know I needed to use `--exit-code`, thank you.

emanuele6 committed:
Update "tests/man.test" and remove superflous whitespace in manual.yml

tests/man.test was generated incorrectly and had some duplicate tests,
the new CI workflow didn't like that.

itchyny commented on .github/workflows/manpage.yml:
```diff
@@ -0,0 +1,34 @@
+          rm -f jq.1.prebuilt tests/man.test
+          make jq.1.prebuilt tests/man.test
+      - name: Make sure man.test is up to date
+        run: git diff --exit-code tests/man.test
```
Why not check jq.1.prebuilt as well?

emanuele6 replied:
Because `jq.1.prebuilt` contains the build date (month and year) on its second line:

```roff
.
.TH "JQ" "1" "July 2023" "" ""
```

So it will start failing next month if we don't update `manual.yml`, but update some other file in `docs/`.

Maybe that date could be removed to allow checking if it is up to date here?

```patch
diff --git a/docs/build_manpage.py b/docs/build_manpage.py
index 4f227b2..15b15a6 100755
--- a/docs/build_manpage.py
+++ b/docs/build_manpage.py
@@ -66,8 +66,7 @@ class RoffWalker(object):
         last_tag = None
         while root is not None:
             if root.tag == 'h1':
-                self.__write_cmd('.TH "JQ" "1" "{}" "" ""'.format(
-                    date.today().strftime('%B %Y')))
+                self.__write_cmd('.TH "JQ" "1" "" "" ""')
                 self.__write_cmd('.SH "NAME"')
                 # TODO: properly parse this
                 self.__write_raw(r'\fBjq\fR \- Command\-line JSON processor' +
```

That date is shown at the bottom of the man page

emanuele6 replied:
I saved the original `jq.1.prebuilt`, and then compared the newly generated one with it ignoring the first two lines using.
```sh
test -s jq.1.prebuilt
diff -- <(tail -n +3 jq.1.old) <(tail -n +3 jq.1.prebuilt)
```

itchyny commented on .github/workflows/manpage.yml:
```diff
@@ -0,0 +1,34 @@
+    paths:
+      - '.github/workflows/manpage.yml'
+      - 'docs/**'
+      - 'tests/test.man'
```
man.test not test.man

nicowilliams commented on tests/man.test:
```diff
@@ -219,904 +219,6 @@ map_values(.+1)
 {"a": 1, "b": 2, "c": 3}
 {"a": 2, "b": 3, "c": 4}
 
-pick(.a, .b.c, .x)
```
I'm surprised to see all these deletions.

emanuele6 replied:
They are all duplicate lines afaict

nicowilliams replied:
Why yes, they are.  LGTM then!

nicowilliams commented:
Feel free to merge this!

itchyny commented on .github/workflows/manpage.yml:
```diff
@@ -0,0 +1,39 @@
+    paths:
+      - '.github/workflows/manpage.yml'
+      - 'docs/**'
+      - 'tests/man.test'
```
Add jq.1.prebuilt too?

emanuele6 replied:
thank you

itchyny reviewed (approved):
LGTM!

nicowilliams commented:
Thanks!
